File: android/app/src/main/java/ai/slashh/runtime/TextFeatures.kt
package ai.slashh.runtime

import kotlin.math.sqrt

/**
 * Shared, deterministic text featurization for the stress text classifier — the Kotlin mirror
 * of `model/text_features.py`. This MUST stay byte-for-byte equivalent to the Python side, or
 * the trained `text_stress.pte` would see different vectors on-device than it trained on (the
 * same single-source-of-truth discipline as [ai.slashh.audio.AudioConfig] vs `audio_config.py`).
 *
 * Contract:
 *   tokenize  : lowercase → replace every char not in `a-z` with a space → split on whitespace,
 *               dropping empties. Unigrams only; no stemming, stopwords, or n-grams.
 *   vectorize : L2-normalized term-count vector over the fixed vocab (token → index).
 */
object TextFeatures {
    private val NON_ALPHA = Regex("[^a-z]+")

    /** Tokenize per the shared contract. */
    fun tokenize(text: String): List<String> =
        text.lowercase().replace(NON_ALPHA, " ").split(' ').filter { it.isNotEmpty() }

    /**
     * Build the token→index map from vocab lines (line i → index i). Drops blank lines
     * defensively; the shipped `text_stress_vocab.txt` has exactly one token per line.
     */
    fun loadVocab(lines: List<String>): Map<String, Int> {
        val m = HashMap<String, Int>(lines.size * 2)
        for ((i, raw) in lines.withIndex()) {
            val tok = raw.trim()
            if (tok.isNotEmpty()) m[tok] = i
        }
        return m
    }

    /**
     * L2-normalized term-count vector, shape [vocabSize]. Returns null when no token is in
     * vocab (norm 0) — there is no learned signal to fuse, so the caller should ignore text.
     */
    fun vectorize(text: String, vocab: Map<String, Int>, vocabSize: Int): FloatArray? {
        val counts = FloatArray(vocabSize)
        var any = false
        for (tok in tokenize(text)) {
            val idx = vocab[tok] ?: continue
            counts[idx] += 1f
            any = true
        }
        if (!any) return null
        var ss = 0.0
        for (c in counts) ss += (c.toDouble() * c.toDouble())
        val norm = sqrt(ss).toFloat()
        if (norm <= 0f) return null
        for (i in counts.indices) counts[i] = counts[i] / norm
        return counts
    }
}

File: model/text_features.py
"""Shared text front-end: tokenize + vocab + vectorize (single source of truth).

This is the *golden* reference the Android/Kotlin side must mirror byte-for-byte.
The exported text ``.pte`` consumes the [1, V] float32 vector produced here; the
``.pte`` + the vocab asset are the contract, so this tokenizer/vectorizer pair is
frozen — any change here must be reflected in Kotlin or parity breaks.

Pipeline (must match Kotlin exactly):
  1. tokenize(text):
       t = text.lower()
       t = re.sub(r"[^a-z]+", " ", t)   # every non a-z char -> a single space
       return [w for w in t.split() if w]
  2. vectorize(tokens, vocab):
       counts = zeros(V)
       for tok in tokens: if tok in vocab: counts[vocab[tok]] += 1
       norm = sqrt(sum(counts**2)); vec = counts/norm if norm>0 else counts
     -> [1, V] float32 (L2-normalized term counts; all-zeros if no in-vocab token)

The vocab asset is one token per line, UTF-8, no header, no blank lines; line
index i (0-based) == feature index i.
"""

from __future__ import annotations

import re
from pathlib import Path

import torch

# --- Vocab asset location (the contract file the Kotlin app loads) ---
VOCAB_PATH = Path("android/app/src/main/assets/text_stress_vocab.txt")

# Matches the Kotlin-side tokenizer exactly: collapse every run of non a-z chars
# (after lowercasing) into a separator, then split on whitespace.
_NON_ALPHA = re.compile(r"[^a-z]+")


def tokenize(text: str) -> list[str]:
    """Lowercase, replace every non a-z char with a space, split, drop empties.

    Unigrams only — no stemming, no stopword removal, no n-grams.
    """
    if text is None:
        return []
    t = text.lower()
    t = _NON_ALPHA.sub(" ", t)
    return [w for w in t.split() if w]


def load_vocab(path: str | Path = VOCAB_PATH) -> dict[str, int]:
    """Load the vocab asset into a ``token -> index`` map.

    One token per line; line index (0-based) is the feature index. No header,
    no blank lines. Returned dict preserves that index == line number.
    """
    path = Path(path)
    vocab: dict[str, int] = {}
    with path.open("r", encoding="utf-8") as f:
        for i, line in enumerate(f):
            token = line.rstrip("\n")
            if token:
                vocab[token] = i
    return vocab


def write_vocab(tokens: list[str], path: str | Path = VOCAB_PATH) -> None:
    """Write the vocab asset: one token per line, UTF-8, no header/blank lines.

    ``tokens`` is the ordered list whose position is the feature index.
    """
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    # Join with newline and add a trailing newline; no blank lines, no header.
    path.write_text("\n".join(tokens) + "\n", encoding="utf-8")


def vectorize(tokens: list[str], vocab: dict[str, int]) -> torch.Tensor:
    """Tokens + vocab -> [1, V] float32 L2-normalized term-count vector.

    counts[idx] += 1 for each in-vocab token; then divide by the L2 norm (unless
    the norm is 0, in which case the all-zeros vector is returned unchanged).
    """
    V = len(vocab)
    counts = torch.zeros(V, dtype=torch.float32)
    for tok in tokens:
        idx = vocab.get(tok)
        if idx is not None:
            counts[idx] += 1.0
    norm = torch.sqrt(torch.sum(counts * counts))
    if norm.item() > 0.0:
        counts = counts / norm
    return counts.reshape(1, V)


def text_to_vector(text: str, vocab: dict[str, int]) -> torch.Tensor:
    """Convenience: raw text -> [1, V] float32 vector (tokenize then vectorize)."""
    return vectorize(tokenize(text), vocab)
